File: Assets/Scripts/Combat/DamageCalc.cs
using System;

namespace IL6
{
    public static class DamageCalc
    {
        public struct Input
        {
            public int Base;
            public int Armor;
            public float CritRoll;     // 0..1, normally rng.Next()
            public float CritChance;   // 0..1
            public float CritMult;
        }

        /// <summary>최소 1 보장. 결정적 (RNG는 호출자가 주입).</summary>
        public static int Compute(in Input input)
        {
            float crit = input.CritRoll < input.CritChance ? input.CritMult : 1f;
            float raw = input.Base * crit - input.Armor;
            return Math.Max(1, (int)Math.Floor(raw));
        }
    }
}

File: src/gameplay/combat/damage-calc.ts
export interface DamageInput {
  base: number;
  armor: number;
  critRoll: number;           // 0..1 (rng.next())
  critChance: number;         // 0..1
  critMult: number;
}

export function calcDamage(input: DamageInput): number {
  const crit = input.critRoll < input.critChance ? input.critMult : 1;
  const raw = input.base * crit - input.armor;
  return Math.max(1, Math.floor(raw));
}
